scipy/optimize/linesearch.py

```diff
@@ -696,15 +695,7 @@ def scalar_search_armijo(phi, phi0, derphi0, c1=1e-4, alpha0=1, amin=0):
     # condition.
 
     while alpha1 > amin:       # we are assuming alpha>0 is a descent direction
-        factor = alpha0**2 * alpha1**2 * (alpha1-alpha0)
-        a = alpha0**2 * (phi_a1 - phi0 - derphi0*alpha1) - \
-            alpha1**2 * (phi_a0 - phi0 - derphi0*alpha0)
-        a = a / factor
-        b = -alpha0**3 * (phi_a1 - phi0 - derphi0*alpha1) + \
-            alpha1**3 * (phi_a0 - phi0 - derphi0*alpha0)
-        b = b / factor
-
-        alpha2 = (-b + np.sqrt(abs(b**2 - 3 * a * derphi0))) / (3.0*a)
+        alpha2 = _cubicmin(0, phi0, derphi0, alpha0, phi_a0, alpha1, phi_a1)
```

Comment:
same here, doesn't look the same. can you explain?

Reply:
Here the None return value actually happens, and the `alpha2` value changes. So unclear if this change is correct and/or fully backwards compatible:
```
import numpy as np


def _cubicmin(a, fa, fpa, b, fb, c, fc):
    """
    Finds the minimizer for a cubic polynomial that goes through the
    points (a,fa), (b,fb), and (c,fc) with derivative at a of fpa.

    If no minimizer can be found return None

    """
    # f(x) = A *(x-a)^3 + B*(x-a)^2 + C*(x-a) + D
    with np.errstate(divide='raise', over='raise', invalid='raise'): 
        try: 
            C = fpa 
            db = b - a 
            dc = c - a 
            denom = (db * dc) ** 2 * (db - dc) 
            d1 = np.empty((2, 2)) 
            d1[0, 0] = dc ** 2 
            d1[0, 1] = -db ** 2 
            d1[1, 0] = -dc ** 3 
            d1[1, 1] = db ** 3 
            [A, B] = np.dot(d1, np.asarray([fb - fa - C * db, 
                                            fc - fa - C * dc]).flatten()) 
            A /= denom 
            B /= denom 
            radical = B * B - 3 * A * C 
            xmin = a + (-B + np.sqrt(radical)) / (3 * A) 
        except ArithmeticError: 
            return None 
    try:
        if not np.isfinite(xmin): 
            return None 
    except TypeError:
        print(xmin)
        raise

    return xmin 


for _ in range(100):
    derphi0 = np.random.rand(1)
    alpha0 = np.random.rand(1)
    alpha1 = np.random.rand(1)
    phi_a0 = np.random.rand(1)
    phi_a1 = np.random.rand(1)
    phi0 = np.random.rand(1)

    factor = alpha0**2 * alpha1**2 * (alpha1-alpha0)
    a = alpha0**2 * (phi_a1 - phi0 - derphi0*alpha1) - \
        alpha1**2 * (phi_a0 - phi0 - derphi0*alpha0)
    a = a / factor
    b = -alpha0**3 * (phi_a1 - phi0 - derphi0*alpha1) + \
        alpha1**3 * (phi_a0 - phi0 - derphi0*alpha0)
    b = b / factor
    alpha2 = (-b + np.sqrt(abs(b**2 - 3 * a * derphi0))) / (3.0*a)

    alpha3 = _cubicmin(0, phi0, derphi0, alpha0, phi_a0, alpha1, phi_a1)
    if alpha3 is None:
        print(alpha1)
        alpha3 = alpha1 / 2.0

    try:
        np.testing.assert_allclose(alpha3, alpha2, atol=0, rtol=5e-14)
    except TypeError:
        # If alpha3 is None, this could happen (added before catching that case above)
        print(alpha2, alpha3)
```